analyzers/src/SonarAnalyzer.Common/AnalysisContext/SonarSematicModelReportingContext.cs

```diff
@@ -0,0 +1,39 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2022 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.AnalysisContext;
+
+public sealed class SonarSematicModelReportingContext : SonarCompilationReportingContextBase<SemanticModelAnalysisContext>
```

Comment:
This needs a lot of unit testing. See `SonarCompilationReportingContextTest` for the boring mechanical stuff and `SonarAnalysisContextTest.Register.cs` for the ugly stuff

Reply:
* SonarCompilationReportingContextTest -> Done
* SonarAnalysisContextTest.Register.cs -> I didn't added `RegisterSemanticModelAction` to `SonarAnalysisContext` yet. It is only added to `SonarCompilationStartAnalysisContext`. Should I add it to `SonarAnalysisContext` too and write the required tests?

Reply:
You should **not** add the other method into production. You should write a similar UT for the method you've written. And it should be rather similar to the existing ones in terms of scaffolding.

I'm not sure how much scaffolding you'll need from the other class, as these tests are scaffolding-heavy. In case it's too much, you can consider putting that test into that "wrong" file.

Reply:
> You should not add the other method into production.

What does this refer to?

I'm not sure why we left out some of the [registration points offered by Roslyn](https://github.com/dotnet/roslyn/blob/main/docs/analyzers/Analyzer%20Actions%20Semantics.md). Is there a reason why I can not RegisterSemanticModelAction, as I can do for Tree, Node, Symbol, and so on?

Reply:
You should not add `RegisterSemanticModelAction` to `SonarAnalysisContext` - we don't need that one (yet).

When we'll need one, we'll add it. Just as you've added `SonarCompilationStartAnalysisContext.RegisterSemanticModelAction` here. As we didn't need the others for quite some years.

Reply:
So I still need to add the (ugly) tests, but in the "wrong" file because we expose the registration on this context only, right?

Reply:
Yes.
In a wrong file only if you need a lot of scaffolding and helper methods from that file. It's up to you to consider how much duplication (if any) will be needed.

Reply:
I added the tests in analyzers/tests/SonarAnalyzer.UnitTest/AnalysisContext/SonarCompilationStartAnalysisContextTest.Register.cs I also tested the other Register... methods on the CompilationStartContext. Let me know if the coverage is sufficient.